src/hotspot/share/classfile/symbolTable.cpp

```diff
@@ -472,18 +460,31 @@ Symbol* SymbolTable::do_add_if_needed(const char* name, int len, uintx hash, boo
   SymbolTableGet stg;
   bool clean_hint = false;
   bool rehash_warning = false;
-  Symbol* sym = nullptr;
   Thread* current = Thread::current();
+  Symbol* sym;
+
+  ResourceMark rm(current);
+  const int alloc_size = Symbol::byte_size(len);
+  u1* u1_buf = NEW_RESOURCE_ARRAY(u1, alloc_size);
+  Symbol* tmp = ::new ((void*)u1_buf) Symbol((const u1*)name, len, (heap && !DumpSharedSpaces) ? 1 : PERM_REFCOUNT);
```

Comment:
There's a version of NEW_RESOURCE_ARRAY that takes the Thread argument, so it doesn't have to call Thread::current()

Reply:
The name of the `heap` parameter to this function should be changed. The node, which contains the Symbol, is always C_HEAP allocated. I would suggest renaming it to `is_permanent`. Same for the other functions that have a `heap` parameter.

Reply:
If the symbol is permanent, it's allocated to an arena not C heap but I agree that the variable name should be changed.

Reply:
> There's a version of NEW_RESOURCE_ARRAY that takes the Thread argument, so it doesn't have to call Thread::current()

I've changed it to:
`u1* u1_buf = NEW_RESOURCE_ARRAY_IN_THREAD(current, u1, alloc_size);`

Reply:
> The name of the `heap` parameter to this function should be changed. The node, which contains the Symbol, is always C_HEAP allocated. I would suggest renaming it to `is_permanent`. Same for the other functions that have a `heap` parameter.

Which other functions did you notice?